packages/flutter_tools/lib/src/commands/daemon.dart

```diff
@@ -237,91 +264,142 @@ class DaemonDomain extends Domain {
   }
 }
 
-/// Return the device matching the deviceId field in the args.
-Future<Device> _getDevice(Daemon daemon, Map<String, dynamic> args) async {
-  if (args == null || args['deviceId'] is! String)
-    throw 'deviceId is required';
-
-  List<Device> devices = await daemon.deviceDomain.getDevices();
-  Device device = devices.firstWhere(
-    (Device device) => device.id == args['deviceId'],
-    orElse: () => null
-  );
-
-  if (device == null)
-    throw "device '${args['deviceId']}' not found";
-
-  return device;
-}
-
 /// This domain responds to methods like [start] and [stop].
 ///
-/// It'll be extended to fire events for when applications start, stop, and
-/// log data.
+/// It fires events for application start, stop, and stdout and stderr.
 class AppDomain extends Domain {
   AppDomain(Daemon daemon) : super(daemon, 'app') {
     registerHandler('start', start);
+    registerHandler('restart', restart);
     registerHandler('stop', stop);
     registerHandler('discover', discover);
   }
 
-  Future<dynamic> start(Map<String, dynamic> args) async {
-    Device device = await _getDevice(daemon, args);
+  static int _nextAppId = 0;
+
+  static String _getNextAppId() => 'app-${_nextAppId++}';
+
+  List<AppInstance> _apps = <AppInstance>[];
+
+  Future<Map<String, dynamic>> start(Map<String, dynamic> args) async {
+    String deviceId = _getArgAsString(args, 'deviceId', required: true);
+    String projectDirectory = _getArgAsString(args, 'projectDirectory', required: true);
+    bool startPaused = _getArgAsBool(args, 'startPaused');
+    // TODO(devoncarew): Use the route param.
+    String route = _getArgAsString(args, 'route'); // ignore: unused_local_variable
+    String mode = _getArgAsString(args, 'mode');
+    String target = _getArgAsString(args, 'target');
+
+    Device device = daemon.deviceDomain._getDevice(deviceId);
+    if (device == null)
+      throw "device '$deviceId' not found";
 
-    if (args['projectDirectory'] is! String)
-      throw "projectDirectory is required";
-    String projectDirectory = args['projectDirectory'];
     if (!FileSystemEntity.isDirectorySync(projectDirectory))
       throw "'$projectDirectory' does not exist";
 
+    BuildMode buildMode = parseBuildMode(mode) ?? BuildMode.debug;
+    DebuggingOptions options;
+
+    if (buildMode == BuildMode.release)
+      options = new DebuggingOptions.disabled(buildMode);
+    else
+      options = new DebuggingOptions.enabled(buildMode, startPaused: startPaused);
```

Comment:
I'd make this a switch so that it's clear that you need to think about profile mode when reading this code.